File: Users/jeong-yeongjae/sub_tree_test/presentation/src/main/java/com/hnh/presentation/logic/classification/PoseClassifierProcessor.kt
package com.hnh.presentation.logic.classification

import android.content.Context
import android.util.Log
import androidx.annotation.WorkerThread
import com.google.mlkit.vision.pose.Pose
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.util.*

class PoseClassifierProcessor @WorkerThread constructor(
    context: Context,
    isStreamMode: Boolean,
    Exercise: Pair<String, Int>
) {
    private val isStreamMode: Boolean
    private var emaSmoothing: EMASmoothing? = null
    private var repCounters: MutableList<RepetitionCounter>? = null
    private var poseClassifier: PoseClassifier? = null
    private var lastRepResult: String? = null

    init {
        //Preconditions.checkState(Looper.myLooper() != Looper.getMainLooper())
        this.isStreamMode = isStreamMode
        if (isStreamMode) {
            emaSmoothing = EMASmoothing()
            repCounters = ArrayList()
            lastRepResult = ""
        }
        type = Exercise.second
        CompanionExercise = Exercise
        loadPoseSamples(context, Exercise)
    }

    private fun loadPoseSamples(context: Context, Exercise: Pair<String, Int>) {
        val poseSamples: MutableList<PoseSample> = ArrayList()
        try {
            val exercise = POSE_SAMPLES_FILE + Exercise.first + ".csv"
            val reader = BufferedReader(
                InputStreamReader(context.assets.open(exercise))
            )
            var csvLine = reader.readLine()
            while (csvLine != null) {
                // If line is not a valid {@link PoseSample}, we'll get null and skip adding to the list.
                val poseSample = PoseSample.getPoseSample(csvLine, ",")
                if (poseSample != null) {
                    poseSamples.add(poseSample)
                }
                csvLine = reader.readLine()
            }
        } catch (e: IOException) {
            Log.e(
                TAG,
                "Error when loading pose samples.\n$e"
            )
        }
        poseClassifier = PoseClassifier(poseSamples, Exercise.second)
        if (isStreamMode) {
            for (className in POSE_CLASSES) {
                repCounters!!.add(RepetitionCounter(className))
            }
        }
    }

    /**
     * Given a new [Pose] input, returns a list of formatted [String]s with Pose
     * classification results.
     *
     *
     * Currently it returns up to 2 strings as following:
     * 0: PoseClass : X reps
     * 1: PoseClass : [0.0-1.0] confidence
     */
    fun resetRepCounter(){
        repCounters!!.clear()
        repCounters!!.add(RepetitionCounter(PUSHUPS_CLASS))
        resetRepCounter = false
    }
    @WorkerThread
    fun getPoseResult(pose: Pose): List<String?> {
        //Preconditions.checkState(Looper.myLooper() != Looper.getMainLooper())
        val result: MutableList<String?> = ArrayList()
        var classification = poseClassifier!!.classify(pose)
        if (resetRepCounter) resetRepCounter()
        // Update {@link RepetitionCounter}s if {@code isStreamMode}.
        if (isStreamMode) {
            // Feed pose to smoothing even if no pose found.
            classification = emaSmoothing!!.getSmoothedResult(classification)

            // Return early without updating repCounter if no pose found.
            if (pose.allPoseLandmarks.isEmpty()) {
                result.add(lastRepResult)
                return result
            }
            for (repCounter in repCounters!!) {
                    repsBefore = repCounter.getNumRepeats()
                    repsAfter = repCounter.addClassificationResult(classification)
                if (repsAfter > repsBefore) {
                    // Play a fun beep when rep counter updates.
//                    val tg = ToneGenerator(AudioManager.STREAM_NOTIFICATION, 100)
//                    tg.startTone(ToneGenerator.TONE_PROP_BEEP)
                    lastRepResult = java.lang.String.format(
                        Locale.US, "%s : %d reps", repCounter.getClassName(), repsAfter
//                        Locale.US, "%d reps",  repsAfter
                    )
                    break
                }
            }
            result.add(lastRepResult)
        }

        // Add maxConfidence class of current frame to result if pose is found.
        if (!pose.allPoseLandmarks.isEmpty()) {
            val maxConfidenceClass: String = classification.getMaxConfidenceClass()
            poseState = maxConfidenceClass
            val maxConfidenceClassResult = java.lang.String.format(
                Locale.US,
//                "%s : %.2f confidence",
                "%s : %.2f con",
                maxConfidenceClass, classification.getClassConfidence(maxConfidenceClass)
                        / poseClassifier!!.confidenceRange()
            )
            result.add(maxConfidenceClassResult)
        }
        return result
    }

    companion object {
        private const val TAG = "PoseClassifierProcessor"

        //        private const val POSE_SAMPLES_FILE = "RunModel/DeadLift_csvs_out.csv"
        private const val POSE_SAMPLES_FILE = "RunModel/"
        var type: Int? = null
        var CompanionExercise: Pair<String, Int>? = null
        var repsBefore = 0
        var repsAfter = 0
        var poseState = ""

        // Specify classes for which we want rep counting.
        // These are the labels in the given {@code POSE_SAMPLES_FILE}. You can set your own class labels
        // for your pose samples.
        private const val PUSHUPS_CLASS = "down"
        private const val SQUATS_CLASS = ""
        private val POSE_CLASSES = arrayOf(
//            PUSHUPS_CLASS, SQUATS_CLASS
            PUSHUPS_CLASS
        )
        var resetRepCounter = false
    }


}

File: iOS/FitGAI/PoseLogic/Classification/PoseClassifierProcessor.swift
//
//  PoseClassifierProcessor.swift
//  FitGAI
//
//  Created by 김기훈 on 2021/12/13.
//

import Foundation
import MLKit

class PoseClassifierProcessor : NSObject {
    static var type: Int? = nil
    static var CompanionExercise: (String, Int)? = nil
    static var poseState = ""
    static var repsAfter = 0
    static var repsBefore = 0
    // Specify classes for which we want rep counting.
    // These are the labels in the given {@code POSE_SAMPLES_FILE}. You can set your own class labels
    // for your pose samples.
    private static let PUSHUPS_CLASS = "down"
    private static let SQUATS_CLASS = ""
    private static let POSE_CLASSES = [//            PUSHUPS_CLASS, SQUATS_CLASS
        PUSHUPS_CLASS
    ]
    
    private let isStreamMode:Bool
    private var emaSmoothing:EMASmoothing?
    private var repCounters:[RepetitionCounter]?
    private var poseClassifier:PoseClassifier?
    private var lastRepResult:String?

    init(_ isStreamMode: Bool, _ Exercise: (String, Int)) {
        self.isStreamMode = isStreamMode
        super.init()
        if isStreamMode {
            emaSmoothing = EMASmoothing()
            repCounters = [RepetitionCounter]()
            lastRepResult = ""
        }
        PoseClassifierProcessor.type = Exercise.1
        PoseClassifierProcessor.CompanionExercise = Exercise
        loadPoseSamples(Exercise: Exercise)
    }
    
    
    private func loadPoseSamples(Exercise: (String, Int)) {
        var poseSamples: [PoseSample] = [PoseSample]()
        
        let exerciseFile = Exercise.0

        var dataString: String? = nil
        do {
            dataString = try String(contentsOfFile: Bundle.main.path(forResource: exerciseFile, ofType: "csv")!, encoding: .utf8)
        } catch {
            HHLog.e(.poseDetect, .no0001,"exerciseFile = \(exerciseFile)")
        }
        
        //let csvData = dataString?.data(using: .utf8)
        let rowData:[String] = dataString!.components(separatedBy: "\n")
        for data in rowData{
            let poseSample = PoseSample.getPoseSample(csvLine:data, separator:",")
            if let poseSample = poseSample {
                poseSamples.append(poseSample)
            }
        }
        poseClassifier = PoseClassifier(poseSamples:poseSamples, type:Exercise.1)
        if (isStreamMode) {
            for className in PoseClassifierProcessor.POSE_CLASSES {
                repCounters!.append(RepetitionCounter(className:className))
            }
        }
    }
    
    func getPoseResult(pose: IOSPose.PoseLandmark) -> [String?] {
        //Preconditions.checkState(Looper.myLooper() != Looper.getMainLooper())
        var result: [String?] = [String?]()
        var classification = poseClassifier!.classify(pose:pose)

        // Update {@link RepetitionCounter}s if {@code isStreamMode}.
        if (isStreamMode) {
            // Feed pose to smoothing even if no pose found.
            classification = emaSmoothing!.getSmoothedResult(classificationResult: classification)

            // Return early without updating repCounter if no pose found.
            if (pose.landmark == nil) {
                result.append(lastRepResult)
                return result
            }
            for repCounter in repCounters! {
                PoseClassifierProcessor.repsBefore = repCounter.numRepeats
                PoseClassifierProcessor.repsAfter = repCounter.addClassificationResult(classificationResult:classification)
                if (PoseClassifierProcessor.repsAfter > PoseClassifierProcessor.repsBefore) {
                    // Play a fun beep when rep counter updates.
//                    val tg = ToneGenerator(AudioManager.STREAM_NOTIFICATION, 100)
//                    tg.startTone(ToneGenerator.TONE_PROP_BEEP)
                    lastRepResult = "\(repCounter.className) : \(PoseClassifierProcessor.repsAfter) reps"
                    break
                }
            }
//            result.append(lastRepResult)
        }

        // Add maxConfidence class of current frame to result if pose is found.
        if (pose.landmark != nil) {
            let maxConfidenceClass: String = classification.getMaxConfidenceClass()
            PoseClassifierProcessor.poseState = maxConfidenceClass
            let confidenceData =
            round(classification.getClassConfidence(className:maxConfidenceClass) / Float(poseClassifier!.confidenceRange()) * 1000)/1000
            let maxConfidenceClassResult = "\(PoseClassifierProcessor.repsAfter) reps \(maxConfidenceClass) : \(confidenceData) con"
            result.append(maxConfidenceClassResult)
        }
        return result
    }
     
}
